### PR TITLE
Fix query network for hetero plugin

### PR DESCRIPTION
**Pull request by olpipi**
### Details:
 - Fixes in queryNetwork:
   - If operation removed after transformation in plugins, mark it as supported
   - Check layer name as well as its fused names, because in some networks layer name is not contained in its fused names

### Tickets:
 - CVS-71762

### CODE REVIEW

File: src/inference/dev_api/cpp_interfaces/interface/ie_iplugin_internal.hpp [hunk 1]
```diff
@@ -335,6 +335,17 @@ class INFERENCE_ENGINE_API_CLASS(IInferencePlugin) : public std::enable_shared_f
...
 
+    /**
+     * @brief Returns set of nodes which were removed after transformation.
+     * If originalNetwork contains node1 and transformedNetwork does not
+     * contains node1 in ops list, node1 will be returned.
+     * @param originalNetwork Original network
+     * @param transformedNetwork Transformed network
+     * @return Set of strings which contains removed node names
+     */
+    std::unordered_set<std::string> GetRemovedNodes(const CNNNetwork& originalNetwork,
```
Comment (nosovmik) [inline src/inference/dev_api/cpp_interfaces/interface/ie_iplugin_internal.hpp:346]: Shall this be used also for private plugins and other third-party plugins, like ARM?

Comment (ilyachur) [reply src/inference/dev_api/cpp_interfaces/interface/ie_iplugin_internal.hpp:346]: It can be a separate activity ONNX Runtime uses Myriad, CPU and GPU

Comment (ilyachur) [inline src/inference/dev_api/cpp_interfaces/interface/ie_iplugin_internal.hpp:346]: Please use `ov::Model` instead of CNNNetwork

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 2]
```diff
@@ -907,19 +907,38 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
         auto ops = clonedNetwork.getFunction()->get_ordered_ops();
         std::unordered_set<std::string> supported;
         std::unordered_set<std::string> unsupported;
+
+        auto layerIsSupported = [&](std::shared_ptr<ngraph::Node> op) {
```
Comment (ilyachur) [inline src/plugins/intel_cpu/src/plugin.cpp:911]:
```suggestion
        auto layerIsSupported = [&](const std::shared_ptr<ngraph::Node>& op) {
```

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 3]
```diff
@@ -907,19 +907,38 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
...
+            std::unique_ptr<MKLDNNNode> ptr;
+            try {
+                ptr.reset(MKLDNNNode::factory().create(op, {mkldnn::engine::kind::cpu, 0}, extensionManager, fake_w_cache));
```
Comment (ilyachur) [inline src/plugins/intel_cpu/src/plugin.cpp:914]: Node creation is a very expensive operation. @dmitry-gorokhov Do you have more light way approach to check that operation is supported?

**Review by ilyachur — COMMENTED** [review]
Could you add tests for these changes?

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 4]
```diff
@@ -971,6 +991,9 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
             }
         }
 
+        auto removedNodeNames = GetRemovedNodes(function, clonnedFunction);
```
Comment (apankratovantonp) [inline src/plugins/intel_cpu/src/plugin.cpp:994]: It should be gotten from fused names vector. If it is missed then the problem is in one of the transformations that missed to add it.

Comment (ilyachur) [reply src/plugins/intel_cpu/src/plugin.cpp:994]: We have an issue that not all transformation saved fused names in this rt section.

**Comment by ilyachur** [comment]
Please fix code style and CI

**Comment by olpipi** [comment]
> Please fix code style and CI

done

File: src/tests/functional/plugin/shared/include/behavior/plugin/hetero_query_network.hpp [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+#include <ie/ie_core.hpp>
+
+using namespace InferenceEngine;
+
+namespace HeteroTests {
+
+class HeteroQueryNetworkTest : public ::testing::TestWithParam<std::string>  {
+public:
+    void RunTest(std::string& deviceName) {
+            ASSERT_GT(deviceName.size(), 0);
```
Comment (ilyachur) [inline src/tests/functional/plugin/shared/include/behavior/plugin/hetero_query_network.hpp:19]: indentation

File: src/tests/functional/plugin/shared/include/behavior/plugin/hetero_query_network.hpp [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+        "Reshape_2/shape/Unsqueeze533",
+        "Reshape_2/shape/Unsqueeze535",
+        "Reshape_2/shape",
+        "Reshape_2/Cast_1",
+        "Reshape_2",
+        "Reshape_2:0"};
+
+    for (auto&& name : checkNames) {
+        EXPECT_NE(result.supportedLayersMap.find(name), result.supportedLayersMap.end());
+    }
```
Comment (ilyachur) [inline src/tests/functional/plugin/shared/include/behavior/plugin/hetero_query_network.hpp:464]: indentation

File: src/plugins/intel_gpu/src/plugin/plugin.cpp [hunk 7]
```diff
@@ -436,34 +439,51 @@ QueryNetworkResult Plugin::QueryNetwork(const CNNNetwork& network,
...
         }
         if (ngraph::is_type<const ngraph::op::v0::Constant>(node)) {
-            constantsNames.emplace(node->get_friendly_name());
             constants.push_back(node);
             return false;
         }
-        return prog.IsOpSupported(network, node) &&
-               !ngraph::op::is_parameter(node) &&
-               !ngraph::op::is_output(node);
+        return prog.IsOpSupported(network, node) ||
```
Comment (ilyachur) [inline src/plugins/intel_gpu/src/plugin/plugin.cpp:445]: Why did you change the logic?

**Review by ilyachur — COMMENTED** [review]
Please resolve merge conflicts

File: src/plugins/intel_gpu/src/plugin/plugin.cpp (hunk 7 above)
Comment (olpipi) [reply src/plugins/intel_gpu/src/plugin/plugin.cpp:445]: Some operations may be fused to parameter or output. 
In this code all parameter and outputs are dropped and some fused operations may be lost here.

There is some additional logic to save all fused names for constants. I kept it. But for parameters and outputs I change logic and mark them as supported w/o any additional checks.

Comment (ilyachur) [reply src/plugins/intel_gpu/src/plugin/plugin.cpp:451]: @vladimir-paramuzov Please take a look

Comment (ilyachur) [reply src/plugins/intel_gpu/src/plugin/plugin.cpp:451]: Additional conformance validation doesn't show any degradation.

File: src/plugins/intel_cpu/src/plugin.cpp [hunk 8]
```diff
@@ -915,28 +915,49 @@ QueryNetworkResult Engine::QueryNetwork(const CNNNetwork& network, const std::ma
...
                                || Config::LPTransformsMode::On == engConfig.lpTransformsMode /* or already enabled */;
         const bool enableSnippets = !(conf.cache_dir.empty() || conf.enableDynamicBatch || (conf.enforceBF16
                 && dnnl::impl::cpu::x64::mayiuse(dnnl::impl::cpu::x64::avx512_core)));
         Transformation(clonedNetwork, enableLPT, enableSnippets, isLegacyAPI());
-        auto ops = clonedNetwork.getFunction()->get_ordered_ops();
-        std::unordered_set<std::string> supported;
+        auto ops = clonnedFunction->get_ordered_ops();
+
+        //Mark removed nodes as supported
+        std::unordered_set<std::string> supported = GetRemovedNodes(function, clonnedFunction);;
```
Comment (ilyachur) [inline src/plugins/intel_cpu/src/plugin.cpp:928]:
```suggestion
        std::unordered_set<std::string> supported = GetRemovedNodes(function, clonnedFunction);
```

File: src/plugins/intel_gpu/src/plugin/plugin.cpp [hunk 9]
```diff
@@ -398,10 +398,14 @@ QueryNetworkResult Plugin::QueryNetwork(const CNNNetwork& network,
     auto clonedNetwork = CloneAndTransformNetwork(network, conf);
     auto func = clonedNetwork.getFunction();
     auto ops = func->get_ordered_ops();
-    std::unordered_set<std::string> supported;
+
+    //Mark removed nodes as supported
+    std::unordered_set<std::string> supported = GetRemovedNodes(function, func);;
```
Comment (ilyachur) [inline src/plugins/intel_gpu/src/plugin/plugin.cpp:403]:
```suggestion
    std::unordered_set<std::string> supported = GetRemovedNodes(function, func);
```

**Review by ilyachur — APPROVED** [review]

**Comment by ilyachur** [comment]
Please fix comment in separate PR
